Resolve #25605 cyclic reference in _LRScheduler

huzecong opened the pull request:
Cyclic reference was introduced in a previous version due to runtime overwriting of the bound method `optimizer.step`. This is now avoided by keeping a weak reference to the optimizer instance.

Credit: https://stackoverflow.com/questions/26157952/why-set-a-bound-method-to-python-object-create-a-circular-reference

huzecong committed:
Resolve #25605 cyclic reference in _LRScheduler

Cyclic reference was introduced in a previous version due to runtime
overwriting of the bound method `optimizer.step`. This is now avoided
by keeping a weak reference to the optimizer instance.

ezyang commented:
Thank you! Do you think there is a way you can conveniently test this (e.g., by setting `__del__` on the optimizer and testing that it is promptly disposed?)

vincentqb commented:
> Thank you! Do you think there is a way you can conveniently test this (e.g., by setting `__del__` on the optimizer and testing that it is promptly disposed?)

like [this](https://github.com/pytorch/pytorch/issues/25605#issue-488840368)?

huzecong commented on torch/optim/lr_scheduler.py:
```diff
@@ -29,15 +29,27 @@ def __init__(self, optimizer, last_epoch=-1):
         # `optimizer.step()`
-        def with_counter(func, opt):
+        def with_counter(method):
+            import weakref
```
Sorry, no, I was just being lazy... I'll move it outside the function.

vincentqb replied:
Cool -- don't worry about it then :)

huzecong commented:
@ezyang I can add a unit test similar to the snippet I included in the issue. Would that do?

ezyang commented:
Yes that would be great, ty

huzecong committed:
Fix: multiple schedulers for same optimizer

When multiple schedulers are constructed for the same optimizer,
`optim.step` is only wrapped once.

huzecong committed:
Add test case that checks for cyclic references

ezyang commented on torch/optim/lr_scheduler.py:
```diff
@@ -29,15 +30,32 @@ def __init__(self, optimizer, last_epoch=-1):
             def wrapper(*args, **kwargs):
-                opt._step_count += 1
-                return func(*args, **kwargs)
+                instance = instance_ref()
```
Is it provable that the optimizer instance is always live at this point? If not maybe we should check the return result of instance...

huzecong replied:
I think it should be, since this function will only be accessible through `optimizer.step()`.

vincentqb commented:
By the way, [this](https://github.com/pytorch/pytorch/issues/17630#issuecomment-525928944) in #17630 and [this](https://github.com/pytorch/pytorch/issues/9942#issuecomment-520638174) in #9942 seems related to this PR. Thoughts?

huzecong commented:
#17630 could be due to having cyclic reference, which prevents optimizer and parameters from garbage collected when `del` is called.

For #9942 store a partial function to a bound method would probably also create a self-reference. TBH I don't really see the need for a partial function here; it's perfectly fine to check `self.mode` and such in `_cmp()`. Also, `_cmp()` can just be a `staticmethod`.

Should I also fix this in the same PR?

vincentqb commented:
> For #9942 store a partial function to a bound method would probably also create a self-reference. TBH I don't really see the need for a partial function here; it's perfectly fine to check `self.mode` and such in `_cmp()`. Also, `_cmp()` can just be a `staticmethod`.

Good point: might as well remove `partial` and let `_cmp()` access `self.*`. Even if it doesn't have a negative performance impact, removing it makes the code more readable :)

> Should I also fix this in the same PR?

I'm leaning for a separate one, but I'm ok either way that is convenient for you, thanks!

ezyang commented:
The test fails, btw

```
Sep 06 20:43:36 ======================================================================
Sep 06 20:43:36 FAIL: test_no_cyclic_references (__main__.TestLRScheduler)
Sep 06 20:43:36 ----------------------------------------------------------------------
Sep 06 20:43:36 Traceback (most recent call last):
Sep 06 20:43:36   File "test_optim.py", line 554, in test_no_cyclic_references
Sep 06 20:43:36     "Optimizer should contain no cyclic references")
Sep 06 20:43:36 AssertionError: False is not true : Optimizer should contain no cyclic references
```

ezyang reviewed (changes requested):
fails its test

huzecong committed:
Fix Python version compatibility in test case

Python `gc` module behavior changed in version 3.7. Prior to 3.7,
when calling `gc.get_referrers` on a local variable in a function, the
current frame would be included in the returned list.

huzecong committed:
Remove cyclic reference in ReduceLROnPlateau

ReduceLROnPlateau would previously store a partial function over a
instance method, which creates a cyclic reference. Refactored the code
to directly use attributes instead of creating a partial function.

huzecong commented:
Sorry for the delay, been busy during the week.

The test error was due to a change in Python version 3.7. Prior to 3.7, `gc.get_referrers` would include the current frame if called on a local variable in a function. I've only been testing on Python 3.7 so I didn't notice this locally. My bad.

I've also addressed #9942 (fix `ReduceLROnPlateau`) in a separate commit.

huzecong committed:
Add back `ReduceLROnPlateau.state_dict`

Didn't realize that `ReduceLROnPlateau` inherited from `object` rather
than `_LRScheduler`... Might leave that change to the future.

ezyang commented:
@pytorchbot rebase this please

Commit:
Merge remote-tracking branch 'origin/master' into HEAD

ezyang commented:
Just waiting on CI

vincentqb reviewed (approved):
Thanks again!

huzecong commented:
Cool! Glad to help!

mcarilli commented:
@huzecong This stuff is really cool and highly instructive to a chronic Cuda programmer/beginner Pythonista.  Is there a reason you chose to grab a manual weakref to the optimizer instance instead of using types.MethodType?  In your [original issue](https://github.com/pytorch/pytorch/issues/25605) you suggested `types.MethodType` resolves the circular reference issue (and it certainly seems simpler than manual reference juggling) but types.MethodType is never mentioned again in either the issue or PR discussion afaict.

Does types.MethodType actually resolve the circular reference issue?  I'm very interested because I'm a frequent (ab)user of it.  I realize that generally, functions are descriptors that become MethodTypes instances whenever they're retrieved and called (as in `optimizer.func()`) but the MethodType instance is a temporary that's only used for the call itself.  Even if the MethodType instance does store a direct non-weak reference to `optimizer`, there's no danger of a long-lived reference to `instance` surviving as a result of the `optimizer.func()` invocation, because the MethodType instance dies immediately.  My point is, since for _typical use_ MethodType instances are short-lived, I could imagine that the implementation of MethodType simply stores a non-weak reference to `optimizer`, in which case
```
def wrapper(self, args....):
    ...
optimizer.step = types.MethodType(wrapper, optimizer)
```
would in fact create a non-weak reference cycle for `optimizer`.  Conversely, it's also possible that MethodType is smart enough to store a weakref to `optimizer`, in which case it _would_ be safe to use, in which case I'm curious why you didn't use it.  I can dig into this some more myself but I'm very interested to hear your take...

huzecong commented:
@mcarilli Sorry if the previous issue confused you, but yes, `types.MethodType` does not solve the circular reference issue because it keeps a non-weak reference to the object. This is also mentioned in the [link](https://stackoverflow.com/questions/26157952/why-set-a-bound-method-to-python-object-create-a-circular-reference) I posted in the PR message, which shows that:
```python
obj = Foo()
obj.func = obj.func
```
saves the `MethodType` instance to the object and that resulted in a circular reference. You can also see this from `gc.get_referrers(obj)` which will list `<bound method func of <Foo object at ...>>` as a referrer of `obj`.

But anyway, this isn't a big problem because Python has mechanisms (albeit more expensive) to deal with circular references. For simple personal scripts, it should be fine as long as the object in question is not holding too much memory.

mcarilli commented:
Makes sense, thanks very much!  My only remaining point of confusion is then:  why did you suggest
> The self-reference could be easily prevented by using the self argument and wrapping the created function with types.MethodType(wrapper, opt).
in the original issue?  Did you also assume (as I did) that MethodType was smart enough to store a weakref to self rather than an ordinary non-weak ref?

huzecong commented:
Yeah, that's pretty much what I thought at the beginning :D
